Fix context stack

rpgoldman opened the pull request:
Previously, it was possible to get the wrong type of `Context` subclass when using `Context.get_context()`.  The patch in the second commit fixes this error by searching the stack for values of the right type.  The first commit gives a test that illustrates the incorrect behavior.

If you wish to verify the error, fetch the branch for this PR, and then checkout the first commit only, and run `pymc3/tests/test_modelcontext.py`

I stumbled on this when vectorizing `sample_posterior_predictive`, because I used `modelcontext()` inside `with _DrawValuesContext():`, which does not happen in the current code base.

Nevertheless, I think that this is worth fixing now, because it's a bug waiting to catch us, and if the bug *does* manifest, then you can get back a `_DrawValuesContext` from a call to `modelcontext(None)`, and this bug is *not* guaranteed to be detected when you try to dereference a name inside what you *think* is a model, but is in fact a `_DrawValuesContext`.  Instead of an error, you just get back `None` when you should have gotten a variable, which is the worst kind of failure.

rpgoldman committed:
Test correct behavior of context stack.

rpgoldman committed:
Check return value type in context stack.

Previously, we simply assumed that the top of the context stack would be of the right type, but that wasn't guaranteed (it just happened to work).
This ensures that we don't return the wrong type from Context.get_contexts().

rpgoldman committed:
Allow caller control of get_context() errors.

Some code expects `Model.get_context()` to error out if the return is
None.  Some code expects `_DrawValuesHandler.get_context()` to return
None.  To harmonize this, added optional argument to control error
signaling.  Defaults to signaling error, because this seems to be the
general case.

rpgoldman committed:
Metaclass-based solution.

Alternative approach to solving the context stack issues.
The previous version was a dead end, because there was no effective way to check the appropriateness of the classes on a single context stack.
So in this version, I split the context stacks into two, one for the pm.Models, and one for pm.distributions.distributions._DrawValuesContext.
This works better, but involved replacing the Context *class* with the ContextMeta parameterized metaclass.

rpgoldman committed:
Fix error in threading.

Initializing the contexts property of the context_class of classes
with ContextMeta as their metaclass was tricky to do properly.

ColCarroll commented:
this looks good to me -- seems like 3.5 is having typing problems

rpgoldman committed:
Fix syntax for 3.5 compatibility.

rpgoldman commented:
> this looks good to me -- seems like 3.5 is having typing problems

Sorry about that -- will I never learn?  If this works, and @lucianopaz likes it, I propose to merge (with a little interactive rebasing to clean up).

ColCarroll commented:
I'll wait for his input - we squash and merge by default, so no big worries about cleaning up the commits.

rpgoldman commented:
> I'll wait for his input - we squash and merge by default, so no big worries about cleaning up the commits.

Squash and merge is fine, but if you want to see the bug I was trying to fix, we might want to squash into one commit introducing the new test, and one introducing the fix.  But if that's too fussy, one squashed commit is fine.

I just think it's important not to put garbage in the history by including my first try at a solution, which didn't work.

lucianopaz commented on pymc3/model.py:
```diff
@@ -161,49 +162,130 @@ def _get_named_nodes_and_relations(graph, parent, leaf_nodes,
+        # you'll get a "TypeError: type() takes 1 or 3 arguments" exception.   
+        return super().__new__(cls, name, bases, dct)
+
+    # FIXME: is there a more elegant way to automatically add methods to the class that
```
Yes, you can add the methods dynamically inside `__new__` by appending them to the `dct`. Check out [this SO question](https://stackoverflow.com/q/32648767/7253166) and [this blog post on how to add them as bound method](https://www.ianlewis.org/en/dynamically-adding-method-classes-or-class-instanc).

rpgoldman replied:
A couple of questions:

1. I can't decide whether it's better to do this in `__new__` or `__init__`, or how I would make that decision.  Is there some reason it would be better to do this in `__new` than in `__init__`, the way I do?
2. Doing this in `__new__` by modifying the dict seems no nicer, syntactically, than what I do.  What I was looking for was a way to do this that would make my intentions clear to a maintainer reading my code afterwards.  Neither of these alternatives seems very nice, because they read as so different from adding a method in a normal class declaration.  In fact, I think messing with the dictionary might be even less readable than the assignment on lines 201 and 202 below.

lucianopaz replied:
I'm not sure if it's better to use `__new__` or `__init__` in our case. I think that using `__new__` should be preferred because the newly instantiated class will have all of its methods available when `__init__` is called. This could be useful if someone ever wants to write a new metaclass that inherits from `ContextMeta`, and shouldn't have to worry about when `supe().__init__(...)` gets called. Of course, I think that this is highly unlike to happen in our case, but I think that given that doing things in `__new__` or `__init__` is almost the same, we can just go for `__new__`.

rpgoldman replied:
OK, I'll make the change, then if the tests pass, I'll merge.  I'm kind of over-extended at work right now, so it might take me a little time.

lucianopaz commented on pymc3/model.py:
```diff
@@ -161,49 +162,130 @@ def _get_named_nodes_and_relations(graph, parent, leaf_nodes,
+
+    # Inherit context class from parent
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
```
Nice!

rpgoldman replied:
Thanks!

lucianopaz commented on pymc3/model.py:
```diff
@@ -161,49 +162,130 @@ def _get_named_nodes_and_relations(graph, parent, leaf_nodes,
     """
     if model is None:
-        return Model.get_context()
+        found = Model.get_context(error_if_none=False) # type: Optional['Model']
```
```suggestion
        model = Model.get_context(error_if_none=False) # type: Optional['Model']
    if model is None:
        raise ValueError("No model on context stack.")
    return model
```
I would dedent and leave a single `return`

rpgoldman replied:
I am making this change and will push it.

lucianopaz reviewed (approved):
Very nice work @rpgoldman! I left a few minor comments. Feel free to ignore them though. The only one that would require some work is to set `__enter__` and `__exit__` as bound methods in the metaclass' `__new__`.

rpgoldman committed:
Adopt Luciano's simplified modelcontext().

rpgoldman commented:
@ColCarroll Now that we have had our release (haven't we?), will you please merge this?

rpgoldman committed:
Typing improvements.

rpgoldman committed:
Fix incorporate_methods().

This function offered a `default` argument which was ignored. Since it was advertised as doing something that it does not do, it's not appropriate to just deprecate it: it must be removed.

rpgoldman committed:
Refactor ContextMeta.

Per @lucianopaz, move the installation of `__enter__` and `__exit__` methods from `__init__()` to `__new__()`.

rpgoldman commented:
If this passes, I've done the fix @lucianopaz requested, so I think this can be merged.

lucianopaz reviewed (approved):
Great work @rpgoldman!
